File: benchmarks/server/src/sirannon_bench/core/rng.py
"""Seeded randomness for reproducible workloads.

Every random choice a workload makes, which key to read, which row to update, what value to
write, comes from one seeded generator, so a run reproduces the same operation stream given the
same seed. The generator wraps Python's Mersenne Twister, which is deterministic across
platforms for an integer seed.
"""

from __future__ import annotations

import random
import string


class SeededRng:
    def __init__(self, seed: int) -> None:
        self._random = random.Random(seed)

    def fraction(self) -> float:
        """Return a float in the half-open interval [0, 1)."""
        return self._random.random()

    def below(self, exclusive_upper: int) -> int:
        """Return an integer in [0, exclusive_upper). Callers guarantee a positive bound."""
        return self._random.randrange(exclusive_upper)

    def text(self, length: int) -> str:
        alphabet = string.ascii_letters + string.digits
        return "".join(self._random.choices(alphabet, k=length))


class ZipfianGenerator:
    """YCSB-compatible Zipfian sampler (Gray et al., SIGMOD 1994).

    The constant ``theta`` fixes the skew; the YCSB default of 0.99 concentrates traffic on a
    few hot keys. The zeta constants are precomputed once so each sample is O(1).
    """

    def __init__(self, items: int, theta: float = 0.99) -> None:
        if items < 1:
            raise ValueError(f"zipfian item count must be positive, got {items}")
        self._items = items
        self._theta = theta
        self._zeta_n = self._compute_zeta(items, theta)
        zeta_2 = self._compute_zeta(2, theta)
        self._alpha = 1.0 / (1.0 - theta)
        self._eta = (1.0 - (2.0 / items) ** (1.0 - theta)) / (1.0 - zeta_2 / self._zeta_n)

    def next(self, rng: SeededRng) -> int:
        u = rng.fraction()
        uz = u * self._zeta_n
        if uz < 1.0:
            return 0
        if uz < 1.0 + 0.5**self._theta:
            return 1
        return min(self._items - 1, int(self._items * (self._eta * u - self._eta + 1.0) ** self._alpha))

    @staticmethod
    def _compute_zeta(n: int, theta: float) -> float:
        total = 0.0
        for i in range(n):
            total += 1.0 / (i + 1) ** theta
        return total

File: benchmarks/server/driver/src/rng.ts
// Seeded randomness so a run reproduces the same operation stream given the same seed. Both
// engines are driven by this one generator, so the two sides answer an identical request stream.
// mulberry32 is a small, fast, fully deterministic PRNG; an integer seed fixes the whole sequence.

const ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

export class SeededRng {
  private state: number

  constructor(seed: number) {
    this.state = seed >>> 0
  }

  fraction(): number {
    this.state = (this.state + 0x6d2b79f5) | 0
    let t = this.state
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }

  below(exclusiveUpper: number): number {
    return Math.floor(this.fraction() * exclusiveUpper)
  }

  text(length: number): string {
    const chars = new Array<string>(length)
    for (let i = 0; i < length; i++) {
      chars[i] = ALPHABET.charAt(Math.floor(this.fraction() * ALPHABET.length))
    }
    return chars.join('')
  }
}

// YCSB-compatible Zipfian sampler (Gray et al., SIGMOD 1994). theta fixes the skew; the YCSB
// default of 0.99 concentrates traffic on a few hot keys. The zeta constants are precomputed once
// so each sample is O(1).
export class ZipfianGenerator {
  private readonly items: number
  private readonly theta: number
  private readonly zetaN: number
  private readonly alpha: number
  private readonly eta: number

  constructor(items: number, theta = 0.99) {
    if (items < 1) {
      throw new Error(`zipfian item count must be positive, got ${items}`)
    }
    this.items = items
    this.theta = theta
    this.zetaN = ZipfianGenerator.computeZeta(items, theta)
    const zeta2 = ZipfianGenerator.computeZeta(2, theta)
    this.alpha = 1.0 / (1.0 - theta)
    this.eta = (1.0 - (2.0 / items) ** (1.0 - theta)) / (1.0 - zeta2 / this.zetaN)
  }

  next(rng: SeededRng): number {
    const u = rng.fraction()
    const uz = u * this.zetaN
    if (uz < 1.0) {
      return 0
    }
    if (uz < 1.0 + 0.5 ** this.theta) {
      return 1
    }
    return Math.min(this.items - 1, Math.trunc(this.items * (this.eta * u - this.eta + 1.0) ** this.alpha))
  }

  private static computeZeta(n: number, theta: number): number {
    let total = 0.0
    for (let i = 0; i < n; i++) {
      total += 1.0 / (i + 1) ** theta
    }
    return total
  }
}
